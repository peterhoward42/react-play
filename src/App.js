/** A very simple DIY example of REACT usage.
 * 
 */

import React, { Component } from 'react';
import './App.css';

const IPSUM = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

const LIST_OF_WORDS = IPSUM.split(/\s/);

/** An element where you can type in the beginning of a word.
 * Emits the characters you enter one at a time.
 * Holds no state of its own - you must provide the text you want
 * to be displayed for each render.
 */
function WordFragmentInput(props) {
  return (
    <div>
      <label>Filter:</label>
      <input
        type="text"
        value={props.textToDisplay}
        placeholder={props.placeHolderText}
        onChange={props.changeSubscriber} />
    </div>
  );
}

/** An element that displays one line of text. */
function WordRow(props) {
  return (<div>{props.word}</div>);
}

/** An element that displays a sequence of lines of text. */
function WordList(props) {
  let counter = 0;
  const listElements = props.listOfWords.map((word) =>
    <WordRow word={word} key={counter++} />
  );
  return <div>{listElements}</div>
}

/** An element that shows a toggle-page link and emits it being
 * clicked.
 */
function TogglePageLink(props) {
  return (<a href='#' onClick={props.clickSubscriber}>Toggle Page</a>);
}

/** Renders the main page, and is responsible for the state and interpretation
 * of word filtering.
 */
class MainPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filterString: ""
    };
    this.handleFilterTextChanged = this.handleFilterTextChanged.bind(this);
  }

  /** A method that interprets the current state of the word filter, to provide
   * a list of filtered words.
   */
  filteredWords() {
    const filterString = this.state.filterString;
    if (filterString === "") {
      return LIST_OF_WORDS;
    }
    return LIST_OF_WORDS.filter((word) => { return word.startsWith(filterString) });
  }

  /** A handler that subscribes to changes to the filter text from a lower level UI
   * component, and actions them.
   */
  handleFilterTextChanged(evt) {
    this.setState({ filterString: evt.target.value });
  }

  render() {
    return (
      <div>
        <WordFragmentInput
          textToDisplay={this.state.filterString}
          changeSubscriber={this.handleFilterTextChanged}
          placeHolderText='Start typing...'
          />
        <WordList listOfWords={this.filteredWords()} />
        <TogglePageLink clickSubscriber={this.props.handleTogglePage} />
      </div>
    );
  }
}

/** A trivial element that can be rendered as a whole page instead of the main one.
 * Exists only to exercise conditional rendering of alternative views.
 * Shows only the link to toggle the view back to the main page.
 */
function AlternativePage(props) {
  return (<a href='#' onClick={props.handleTogglePage}>Toggle Page</a>);
}

/** The top level element that aggregates the choice of two pages and owns
 * the state for which page is selected.
  */
class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      onMainPage: true
    };
    this.handleTogglePage = this.handleTogglePage.bind(this);
  }

  handleTogglePage(evt) {
    const toggle = !this.state.onMainPage;
    this.setState({ onMainPage: toggle });
  }

  /** The entry point render method. */
  render() {
    if (this.state.onMainPage) {
      return (
        <MainPage handleTogglePage={this.handleTogglePage} />
      );
    }
    else {
      return (<AlternativePage handleTogglePage={this.handleTogglePage} />);
    }
  }
}

export default App;
