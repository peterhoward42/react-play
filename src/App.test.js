// Illustrates some examples of testing using Jest

// See also v good tutorial on TDD for react using enzyme here:
// http://thereignn.ghost.io/a-step-by-step-tdd-approach-on-testing-react-components-using-enzyme/

import React from 'react';
import ReactDOM from 'react-dom';
import { App, MainPage, AlternativePage, WordRow } from './App';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';


// Start with plan js testing, i.e. not react specific

describe('Testing plain js', () => {
  it('can divide', function () {
    expect(12 / 4).toEqual(3)
  })
  it('can square', function () {
    expect(8 * 8).toEqual(64)
  })
})


describe('Exercise a class method', () => {
  it('filters the words correctly', function () {
    let mainPage = new MainPage()
    mainPage.state = {
      filterString: "s"
    };
    const words = mainPage.filteredWords()
    const sampledWord = words[3];
    expect(sampledWord).toEqual('sunt')
  })
})

// Play around with expect (jest provides its own expect function)

it('The expectation object provides a mirror set of inverted tests', () => {
  expect(42).toBe(42)
  expect(99).not.toBe(42)
})

it('Tries out the toHaveLength test', () => {
  expect('012345789').toHaveLength(9)
})

it('Tries out toBeInstanceOf', function () {
  let mainPage = new MainPage()
  expect(mainPage).toBeInstanceOf(MainPage)
})

it('Tries out (mock function) toHaveBeenCalled', function () {
  // Meaningless with explicit call like this, but used when the
  // a function is passed into something as a callback.
  let mockFn = jest.fn() // A mock function
  mockFn()
  expect(mockFn).toHaveBeenCalled()
})

it('Tries out (mock function) toHaveBeenCalledWith', function () {
  // Similar to toHaveBeenCalled, but checks a call happened with arguments>
  // equal to those given (=== equality).
  let mockFn = jest.fn()
  const objectA = { count: 3 }
  const sameAsObjectA = { count: 3 }
  mockFn(objectA)
  expect(mockFn).toHaveBeenCalledWith(sameAsObjectA)
})

// These full rendering tests I think are let you inspect the DOM created in
// depth and detail.
it('Exercises ReactDOM.render() to show a standard HTML element can be rendered without crashing', () => {
  const wrapper = document.createElement('div');
  const element = <p>foo</p>
  ReactDOM.render(element, wrapper);
})

// These shallow rendering tests I think are let you inspect the DOM created in
// depth and detail.
it('Exercises ReactDOM.render() to show a standard HTML element can be rendered without crashing', () => {
  const wrapper = document.createElement('div');
  const element = <p>foo</p>
  ReactDOM.render(element, wrapper);
})

it('Exercises shallow.render() to render a component without needing a context, to show that it is not dependent on its children', () => {
  const wrapper = shallow(<App />)
  expect(wrapper.find(MainPage)).toHaveLength(1)
})

// Basic snapshot test demo

it('Shows a very simple example of snapshot', () => {
  const tree = renderer.create(
    <App/>
  ).toJSON() // seems this version of toJSON() gives the tree some functions too?
  expect(tree).toMatchSnapshot()
  // stimulate change in state - tutorial says to do this by calling tree.props but cannot see how tree which is json can have a method, surely they maean component not tree?
  expect(tree).toMatchSnapshot()
})





