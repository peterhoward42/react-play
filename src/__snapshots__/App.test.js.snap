exports[`test Shows a very simple example of snapshot 1`] = `
<div>
  <div>
    <label>
      Filter:
    </label>
    <input
      onChange={[Function]}
      placeholder="Start typing..."
      type="text"
      value="" />
  </div>
  <div>
    <div>
      Lorem
    </div>
    <div>
      ipsum
    </div>
    <div>
      dolor
    </div>
    <div>
      sit
    </div>
    <div>
      amet,
    </div>
    <div>
      consectetur
    </div>
    <div>
      adipiscing
    </div>
    <div>
      elit,
    </div>
    <div>
      sed
    </div>
    <div>
      do
    </div>
    <div>
      eiusmod
    </div>
    <div>
      tempor
    </div>
    <div>
      incididunt
    </div>
    <div>
      ut
    </div>
    <div>
      labore
    </div>
    <div>
      et
    </div>
    <div>
      dolore
    </div>
    <div>
      magna
    </div>
    <div>
      aliqua.
    </div>
    <div>
      Ut
    </div>
    <div>
      enim
    </div>
    <div>
      ad
    </div>
    <div>
      minim
    </div>
    <div>
      veniam,
    </div>
    <div>
      quis
    </div>
    <div>
      nostrud
    </div>
    <div>
      exercitation
    </div>
    <div>
      ullamco
    </div>
    <div>
      laboris
    </div>
    <div>
      nisi
    </div>
    <div>
      ut
    </div>
    <div>
      aliquip
    </div>
    <div>
      ex
    </div>
    <div>
      ea
    </div>
    <div>
      commodo
    </div>
    <div>
      consequat.
    </div>
    <div>
      Duis
    </div>
    <div>
      aute
    </div>
    <div>
      irure
    </div>
    <div>
      dolor
    </div>
    <div>
      in
    </div>
    <div>
      reprehenderit
    </div>
    <div>
      in
    </div>
    <div>
      voluptate
    </div>
    <div>
      velit
    </div>
    <div>
      esse
    </div>
    <div>
      cillum
    </div>
    <div>
      dolore
    </div>
    <div>
      eu
    </div>
    <div>
      fugiat
    </div>
    <div>
      nulla
    </div>
    <div>
      pariatur.
    </div>
    <div>
      Excepteur
    </div>
    <div>
      sint
    </div>
    <div>
      occaecat
    </div>
    <div>
      cupidatat
    </div>
    <div>
      non
    </div>
    <div>
      proident,
    </div>
    <div>
      sunt
    </div>
    <div>
      in
    </div>
    <div>
      culpa
    </div>
    <div>
      qui
    </div>
    <div>
      officia
    </div>
    <div>
      deserunt
    </div>
    <div>
      mollit
    </div>
    <div>
      anim
    </div>
    <div>
      id
    </div>
    <div>
      est
    </div>
    <div>
      laborum.
    </div>
  </div>
  <a
    href="#"
    onClick={[Function]}>
    Toggle Page
  </a>
</div>
`;

exports[`test Shows a very simple example of snapshot 2`] = `
<div>
  <div>
    <label>
      Filter:
    </label>
    <input
      onChange={[Function]}
      placeholder="Start typing..."
      type="text"
      value="" />
  </div>
  <div>
    <div>
      Lorem
    </div>
    <div>
      ipsum
    </div>
    <div>
      dolor
    </div>
    <div>
      sit
    </div>
    <div>
      amet,
    </div>
    <div>
      consectetur
    </div>
    <div>
      adipiscing
    </div>
    <div>
      elit,
    </div>
    <div>
      sed
    </div>
    <div>
      do
    </div>
    <div>
      eiusmod
    </div>
    <div>
      tempor
    </div>
    <div>
      incididunt
    </div>
    <div>
      ut
    </div>
    <div>
      labore
    </div>
    <div>
      et
    </div>
    <div>
      dolore
    </div>
    <div>
      magna
    </div>
    <div>
      aliqua.
    </div>
    <div>
      Ut
    </div>
    <div>
      enim
    </div>
    <div>
      ad
    </div>
    <div>
      minim
    </div>
    <div>
      veniam,
    </div>
    <div>
      quis
    </div>
    <div>
      nostrud
    </div>
    <div>
      exercitation
    </div>
    <div>
      ullamco
    </div>
    <div>
      laboris
    </div>
    <div>
      nisi
    </div>
    <div>
      ut
    </div>
    <div>
      aliquip
    </div>
    <div>
      ex
    </div>
    <div>
      ea
    </div>
    <div>
      commodo
    </div>
    <div>
      consequat.
    </div>
    <div>
      Duis
    </div>
    <div>
      aute
    </div>
    <div>
      irure
    </div>
    <div>
      dolor
    </div>
    <div>
      in
    </div>
    <div>
      reprehenderit
    </div>
    <div>
      in
    </div>
    <div>
      voluptate
    </div>
    <div>
      velit
    </div>
    <div>
      esse
    </div>
    <div>
      cillum
    </div>
    <div>
      dolore
    </div>
    <div>
      eu
    </div>
    <div>
      fugiat
    </div>
    <div>
      nulla
    </div>
    <div>
      pariatur.
    </div>
    <div>
      Excepteur
    </div>
    <div>
      sint
    </div>
    <div>
      occaecat
    </div>
    <div>
      cupidatat
    </div>
    <div>
      non
    </div>
    <div>
      proident,
    </div>
    <div>
      sunt
    </div>
    <div>
      in
    </div>
    <div>
      culpa
    </div>
    <div>
      qui
    </div>
    <div>
      officia
    </div>
    <div>
      deserunt
    </div>
    <div>
      mollit
    </div>
    <div>
      anim
    </div>
    <div>
      id
    </div>
    <div>
      est
    </div>
    <div>
      laborum.
    </div>
  </div>
  <a
    href="#"
    onClick={[Function]}>
    Toggle Page
  </a>
</div>
`;
